pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -1075,10 +1075,23 @@ public CompletableFuture<Optional<Topic>> getTopic(final TopicName topicName, bo
                                             return loadOrCreatePersistentTopic(tpName, createIfMissing,
                                                     properties, topicPolicies);
                                         }
-                                        return CompletableFuture.completedFuture(Optional.empty());
+                                        final String info =
+                                                String.format("Creating a topic encountered an illegal partition name."
+                                                                + " topic_name=%s metadata_partition_number=%s",
+                                                        topicName, metadata.partitions);
+                                        log.warn(info);
+                                        return CompletableFuture
+                                                .failedFuture(new BrokerServiceException.NamingException(info));
                                     });
                         }
                         return loadOrCreatePersistentTopic(tpName, createIfMissing, properties, topicPolicies);
+                    }).thenCompose(optionalTopic -> {
```

Comment:
Can we change `optionalTopic` -> `Pair<isValidTopic, optionalTopic>` so that we don't need to return a failed future at https://github.com/apache/pulsar/pull/21545/files#diff-0210356c8a88e4efa89eb769a027fa6c166db479dbad8bbbbc704c6ed6e317f5R1083

Then we should only retry to load the topic when `if(isValidTopic && optionalTopic.isEmptry() && createIfMissing)`

Following this pattern, we will not change any behavior related to the invalid partition name. If it also makes sense to improve the invalid partition name, it should be happened in another PR.

Reply:
Sorry, the above solution will not work since the future is created in the map.computeIfAbsent.


```diff
--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java
@@ -90,6 +90,7 @@ import org.apache.bookkeeper.mledger.impl.NullLedgerOffloader;
 import org.apache.bookkeeper.mledger.util.Futures;
 import org.apache.commons.collections4.MapUtils;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.pulsar.bookie.rackawareness.IsolatedBookieEnsemblePlacementPolicy;
 import org.apache.pulsar.broker.PulsarServerException;
 import org.apache.pulsar.broker.PulsarService;
@@ -1064,34 +1065,31 @@ public class BrokerService implements Closeable {
                     throw FutureUtil.wrapToCompletionException(new ServiceUnitNotReadyException(errorInfo));
                 }).thenCompose(optionalTopicPolicies -> {
                     final TopicPolicies topicPolicies = optionalTopicPolicies.orElse(null);
-                    return topics.computeIfAbsent(topicName.toString(), (tpName) -> {
-                        if (topicName.isPartitioned()) {
-                            final TopicName topicNameEntity = TopicName.get(topicName.getPartitionedTopicName());
-                            return fetchPartitionedTopicMetadataAsync(topicNameEntity)
-                                    .thenCompose((metadata) -> {
-                                        // Allow crate non-partitioned persistent topic that name includes `partition`
-                                        if (metadata.partitions == 0
-                                                || topicName.getPartitionIndex() < metadata.partitions) {
-                                            return loadOrCreatePersistentTopic(tpName, createIfMissing,
-                                                    properties, topicPolicies);
-                                        }
-                                        final String info =
-                                                String.format("Creating a topic encountered an illegal partition name."
-                                                                + " topic_name=%s metadata_partition_number=%s",
-                                                        topicName, metadata.partitions);
-                                        log.warn(info);
-                                        return CompletableFuture
-                                                .failedFuture(new BrokerServiceException.NamingException(info));
-                                    });
-                        }
-                        return loadOrCreatePersistentTopic(tpName, createIfMissing, properties, topicPolicies);
-                    }).thenCompose(optionalTopic -> {
-                        if (!optionalTopic.isPresent() && createIfMissing) {
-                            log.warn("Different topic automatic creation strategies lead to race conditions. "
-                                    + "Try again to try to recover. topic_name={}", topicName);
-                            return getTopic(topicName, true, properties);
+                    CompletableFuture<Boolean> isValidTopic = CompletableFuture.completedFuture(true);
+                    if (topicName.isPartitioned()) {
+                        final TopicName topicNameEntity = TopicName.get(topicName.getPartitionedTopicName());
+                        isValidTopic = fetchPartitionedTopicMetadataAsync(topicNameEntity)
+                                .thenApply((metadata) -> {
+                                    // Allow crate non-partitioned persistent topic that name includes `partition`
+                                    return metadata.partitions == 0
+                                            || topicName.getPartitionIndex() < metadata.partitions;
+                                });
+                    }
+                    return isValidTopic.thenCompose(isValid -> {
+                        if (isValid) {
+                            return topics.computeIfAbsent(topicName.toString(), (tpName) -> {
+                                return loadOrCreatePersistentTopic(tpName, createIfMissing, properties, topicPolicies);
+                            }).thenCompose(optionalTopic -> {
+                                if (!optionalTopic.isPresent() && createIfMissing) {
+                                    log.warn("Different topic automatic creation strategies lead to race conditions. "
+                                            + "Try again to try to recover. topic_name={}", topicName);
+                                    return getTopic(topicName, true, properties);
+                                }
+                                return CompletableFuture.completedFuture(optionalTopic);
+                            });
+                        } else {
+                            return CompletableFuture.completedFuture(Optional.empty());
                         }
-                        return CompletableFuture.completedFuture(optionalTopic);
                     });
                 });
             } else {
```

It looks like we don't need to check if the topic name is a valid topic name in the map.computeIfAbsent? So I tried the above changes. @mattisonchao

Reply:
This approach has side effects on performance. :)